src/AppInstallerRepositoryCore/RepositorySource.cpp

```diff
@@ -964,31 +951,44 @@ namespace AppInstaller::Repository
             {
                 AICLI_LOG(Repo, Info, << "Named source requested, found: " << source->Name);
 
-                OpenSourceResult result;
+                OpenSourceResult result = OpenSourceFromDetails(*source, progress);
+                return result;
+            }
+        }
+    }
 
-                if (ShouldUpdateBeforeOpen(*source))
+    OpenSourceResult OpenSourceFromDetails(SourceDetails& details, IProgressCallback& progress)
+    {
+        SourceListInternal sourceList;
+        OpenSourceResult result;
+
+        if (ShouldUpdateBeforeOpen(details))
+        {
+            try
+            {
+                if (BackgroundUpdateSourceFromDetails(details, progress))
                 {
-                    try
-                    {
-                        if (BackgroundUpdateSourceFromDetails(*source, progress))
-                        {
-                            sourceList.SaveMetadata();
-                        }
-                    }
-                    catch (...)
-                    {
-                        AICLI_LOG(Repo, Warning, << "Failed to update source: " << (*source).Name);
-                        result.SourcesWithUpdateFailure.emplace_back(*source);
-                    }
+                    sourceList.SaveMetadata();

```

Comment:
This doesn't do anything of value; the details passed in may be updated but they will have no effect on the source list.

I hadn't thought of this aspect of things, and thus my previous suggestions may have been a little off of what we actually need.

It might be possible to prevent a rewrite of the whole details/metadata design by having the metadata saved for all details, including tombstones.  But this function would have to first look up the real details from the `SourceListInternal` and use those to update/open.

I actually prefer instead to go with a Predefined model, where you can get the details but still have to open them by enum.  I realize that makes things more complex on the COM method, but I foresee issues with someone creating some arbitrary source details and then wanting to open them and it not working correctly.

If not one of those two, then I think it is a rewrite of the details system to be more object oriented.  Basically making it more like the reference system in the COM API.  It would certainly make the COM API simpler, but I don't think it should be done as part of this change.

For now I would suggest just matching the details with the right value in the source list and using that.  And comment about matching the tombstones.  I'm working on a refactor of the metadata code anyway, I might be able to also move to the OO model at the same time.